File: ts/src/primitive.ts
import { Event } from "./event.js";
import { InvalidTransitionError } from "./errors.js";
import {
  Activation, Cadence, EventId, EventType, Layer, PrimitiveId, SubscriptionPattern,
} from "./types.js";
import type { ActorId, ConversationId } from "./types.js";

// ── Lifecycle ────────────────────────────────────────────────────────────

export const Lifecycle = {
  Dormant: "dormant",
  Activating: "activating",
  Active: "active",
  Processing: "processing",
  Emitting: "emitting",
  Suspending: "suspending",
  Suspended: "suspended",
  Memorial: "memorial",
} as const;

const TRANSITIONS: Record<string, Set<string>> = {
  [Lifecycle.Dormant]: new Set([Lifecycle.Activating]),
  [Lifecycle.Activating]: new Set([Lifecycle.Active]),
  [Lifecycle.Active]: new Set([Lifecycle.Processing, Lifecycle.Suspending, Lifecycle.Memorial]),
  [Lifecycle.Processing]: new Set([Lifecycle.Emitting, Lifecycle.Active]),
  [Lifecycle.Emitting]: new Set([Lifecycle.Active]),
  [Lifecycle.Suspending]: new Set([Lifecycle.Suspended]),
  [Lifecycle.Suspended]: new Set([Lifecycle.Activating, Lifecycle.Memorial]),
  [Lifecycle.Memorial]: new Set(),
};

export function isValidTransition(from: string, to: string): boolean {
  return TRANSITIONS[from]?.has(to) ?? false;
}

// ── Mutations ───────────────────────────────────────────────────────────

export type Mutation =
  | { kind: "addEvent"; type: EventType; source: ActorId; content: Record<string, unknown>; causes: EventId[]; conversationId: ConversationId }
  | { kind: "updateState"; primitiveId: PrimitiveId; key: string; value: unknown }
  | { kind: "updateActivation"; primitiveId: PrimitiveId; level: Activation }
  | { kind: "updateLifecycle"; primitiveId: PrimitiveId; state: string };

// ── Snapshot ─────────────────────────────────────────────────────────────

export interface PrimitiveState {
  id: PrimitiveId;
  layer: Layer;
  lifecycle: string;
  activation: Activation;
  cadence: Cadence;
  state: Record<string, unknown>;
  lastTick: number;
}

export interface Snapshot {
  tick: number;
  primitives: Map<string, PrimitiveState>;
  pendingEvents: Event[];
  recentEvents: Event[];
}

// ── Primitive interface ─────────────────────────────────────────────────

export interface Primitive {
  id(): PrimitiveId;
  layer(): Layer;
  process(tick: number, events: Event[], snapshot: Snapshot): Mutation[];
  subscriptions(): SubscriptionPattern[];
  cadence(): Cadence;
}

// ── Registry ─────────────────────────────────────────────────────────────

interface MutableState {
  activation: Activation;
  lifecycle: string;
  state: Record<string, unknown>;
  lastTick: number;
}

export class Registry {
  private readonly primitives = new Map<string, Primitive>();
  private readonly states = new Map<string, MutableState>();
  private ordered: string[] = [];

  register(p: Primitive): void {
    const key = p.id().value;
    if (this.primitives.has(key))
      throw new Error(`Primitive '${key}' already registered`);
    this.primitives.set(key, p);
    this.states.set(key, {
      activation: new Activation(0),
      lifecycle: Lifecycle.Dormant,
      state: {},
      lastTick: 0,
    });
    this.rebuildOrder();
  }

  get(id: PrimitiveId): Primitive | undefined {
    return this.primitives.get(id.value);
  }

  all(): Primitive[] {
    return this.ordered.map((k) => this.primitives.get(k)!);
  }

  get count(): number { return this.primitives.size; }

  allStates(): Map<string, PrimitiveState> {
    const result = new Map<string, PrimitiveState>();
    for (const [key, p] of this.primitives) {
      const ms = this.states.get(key)!;
      result.set(key, {
        id: p.id(), layer: p.layer(), lifecycle: ms.lifecycle,
        activation: ms.activation, cadence: p.cadence(),
        state: { ...ms.state }, lastTick: ms.lastTick,
      });
    }
    return result;
  }

  getLifecycle(id: PrimitiveId): string {
    return this.states.get(id.value)?.lifecycle ?? Lifecycle.Dormant;
  }

  setLifecycle(id: PrimitiveId, state: string): void {
    const ms = this.states.get(id.value);
    if (!ms) throw new Error(`Primitive '${id.value}' not found`);
    if (!isValidTransition(ms.lifecycle, state))
      throw new InvalidTransitionError(ms.lifecycle, state);
    ms.lifecycle = state;
  }

  activate(id: PrimitiveId): void {
    const ms = this.states.get(id.value);
    if (!ms) throw new Error(`Primitive '${id.value}' not found`);
    if (!isValidTransition(ms.lifecycle, Lifecycle.Activating))
      throw new InvalidTransitionError(ms.lifecycle, Lifecycle.Activating);
    ms.lifecycle = Lifecycle.Active;
  }

  setActivation(id: PrimitiveId, level: Activation): void {
    const ms = this.states.get(id.value);
    if (!ms) throw new Error(`Primitive '${id.value}' not found`);
    ms.activation = level;
  }

  updateState(id: PrimitiveId, key: string, value: unknown): void {
    const ms = this.states.get(id.value);
    if (!ms) throw new Error(`Primitive '${id.value}' not found`);
    ms.state[key] = structuredClone(value);
  }

  getLastTick(id: PrimitiveId): number {
    return this.states.get(id.value)?.lastTick ?? 0;
  }

  setLastTick(id: PrimitiveId, tick: number): void {
    const ms = this.states.get(id.value);
    if (ms) ms.lastTick = tick;
  }

  private rebuildOrder(): void {
    this.ordered = [...this.primitives.keys()].sort((a, b) => {
      const la = this.primitives.get(a)!.layer().value;
      const lb = this.primitives.get(b)!.layer().value;
      return la !== lb ? la - lb : a.localeCompare(b);
    });
  }
}

File: go/pkg/primitive/primitive.go
// Package primitive defines the Primitive interface, Mutation types,
// Snapshot, PrimitiveState, and PrimitiveRegistry. These are the building
// blocks that the tick engine processes.
package primitive

import (
	"github.com/transpara-ai/eventgraph/go/pkg/actor"
	"github.com/transpara-ai/eventgraph/go/pkg/event"
	"github.com/transpara-ai/eventgraph/go/pkg/types"
)

// Primitive is a software agent that embodies a specific domain of intelligence.
// Each primitive has identity, state, lifecycle, subscriptions, and cadence.
// The tick engine invokes Process() when matching events exist and cadence allows.
type Primitive interface {
	ID() types.PrimitiveID
	Layer() types.Layer
	Process(tick types.Tick, events []event.Event, snapshot Snapshot) ([]Mutation, error)
	Subscriptions() []types.SubscriptionPattern
	Cadence() types.Cadence
	Lifecycle() types.LifecycleState
}

// PrimitiveState is the read-only view of a primitive's state used in snapshots.
// The state map is unexported to enforce the Frozen<Snapshot> invariant —
// callers receive a defensive copy via the State() method.
type PrimitiveState struct {
	ID         types.PrimitiveID
	Layer      types.Layer
	Lifecycle  types.LifecycleState
	Activation types.Activation
	Cadence    types.Cadence
	state      map[string]any
	LastTick   types.Tick
}

// State returns a deep defensive copy of the primitive's key-value state.
// Returns nil if the state is nil, preserving nil/empty distinction.
func (ps PrimitiveState) State() map[string]any {
	return deepCopyState(ps.state)
}

// Snapshot is the frozen, read-only view passed to primitives during processing.
// No primitive can mutate another's state through the snapshot.
type Snapshot struct {
	Tick          types.Tick
	Primitives    map[types.PrimitiveID]PrimitiveState
	PendingEvents []event.Event
	RecentEvents  []event.Event
	ActiveActors  []actor.IActor
}

// --- Mutation types ---

// Mutation is a declarative change produced by a primitive's Process() call.
// The tick engine collects and applies mutations atomically after each wave.
type Mutation interface {
	Accept(MutationVisitor)
	isMutation()
}

// MutationVisitor provides exhaustive dispatch over mutation types.
type MutationVisitor interface {
	VisitAddEvent(AddEvent)
	VisitAddEdge(AddEdge)
	VisitUpdateState(UpdateState)
	VisitUpdateActivation(UpdateActivation)
	VisitUpdateLifecycle(UpdateLifecycle)
}

// AddEvent requests the tick engine to create and append a new event.
type AddEvent struct {
	Type           types.EventType
	Source         types.ActorID
	Content        event.EventContent
	Causes         []types.EventID
	ConversationID types.ConversationID // zero value uses engine default
}

func (m AddEvent) Accept(v MutationVisitor) { v.VisitAddEvent(m) }
func (m AddEvent) isMutation()              {}

// AddEdge requests the tick engine to create a new edge (via an edge.created event).
type AddEdge struct {
	From     types.ActorID
	To       types.ActorID
	EdgeType event.EdgeType
	Weight   types.Weight
	Scope    types.Option[types.DomainScope]
	Causes   []types.EventID // optional — if empty, engine uses chain head
}

func (m AddEdge) Accept(v MutationVisitor) { v.VisitAddEdge(m) }
func (m AddEdge) isMutation()              {}

// UpdateState updates a key-value pair in the primitive's internal state.
type UpdateState struct {
	PrimitiveID types.PrimitiveID
	Key         string
	Value       any
}

func (m UpdateState) Accept(v MutationVisitor) { v.VisitUpdateState(m) }
func (m UpdateState) isMutation()              {}

// UpdateActivation changes the primitive's activation level.
type UpdateActivation struct {
	PrimitiveID types.PrimitiveID
	Level       types.Activation
}

func (m UpdateActivation) Accept(v MutationVisitor) { v.VisitUpdateActivation(m) }
func (m UpdateActivation) isMutation()              {}

// UpdateLifecycle requests a lifecycle state transition. Must be a valid transition.
type UpdateLifecycle struct {
	PrimitiveID types.PrimitiveID
	State       types.LifecycleState
}

func (m UpdateLifecycle) Accept(v MutationVisitor) { v.VisitUpdateLifecycle(m) }
func (m UpdateLifecycle) isMutation()              {}
